### PR TITLE
Special Case Error Handler For Unparsable Request Line

### LINKED ISSUE #6931: Ember server responds with 500 when sent illegal characters in the query parameters
I do not know if this is an issue with ember server or http4s core, but if you send characters that should be URL encoded as query params the route that should match never gets called and the server returns 500.

Steps:
1: Clone the [g8 template](https://github.com/http4s/http4s.g8) - I did scala 2.
2: Start the server using `sbt run`
3: Use a browser (or JS) to send an invalid character.
```
await fetch("http://localhost:8080/hello/colt?foo={");
```

Note that the server fails with 500, instead of matching the route and returning the "Hello, colt" response.

This is a problem because people are sending garbage requests to a service I have exposed on the internet and it causes 500s to show up in our logs.

### PR DESCRIPTION
**Pull request by isomarcte**
This commit adds a special case error handler to be executed when the request-line, e.g. "GET / HTTP/1.1", it not able to be parsed.

This is a different case than the more general error handler we already have. RFC 7230 states a server should return either 400 or 301. The handler is configurable, but the default is to return a 301

Closes #6931

### CODE REVIEW

**Comment by isomarcte** [comment]
@coltfred

**Comment by isomarcte** [comment]
I've made this a draft, as I need to add a test still.

File: ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala [hunk 1]
```diff
@@ -485,8 +494,12 @@ private[server] object ServerHelpers extends ServerHelpersPlatform {
                   EmberException.ReadTimeout(_) =>
                 Applicative[F].pure(None)
               case err =>
-                errorHandler(err)
-                  .handleError(_ => serverFailure.covary[F])
+                (err match {
+                  case err: Parser.Request.ReqPrelude.ParsePreludeError =>
+                    requestLineParseErrorHandler(err)
```
Comment (armanbilge) [inline ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:499]: Might be dumb question: doesn't the request prelude involve more than the request line?

Comment (isomarcte) [reply ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:499]: https://github.com/http4s/http4s/blob/6931-special-case-handle-request-line-parse-failure/ember-core/shared/src/main/scala/org/http4s/ember/core/Parser.scala#L198

I don't think so. Method, Http Version, URI.

Comment (armanbilge) [reply ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala:499]: Ah, thanks. So then maybe `ReqPrelude` should be `ReqLine` if that's the "official" name?

**Comment by rossabaker** [comment]
I would also [close the connection](https://www.rfc-editor.org/rfc/rfc9112#section-2.2-9):

> When a server listening only for HTTP request messages, or processing what appears from the start-line to be an HTTP request message, receives a sequence of octets that does not match the HTTP-message grammar aside from the robustness exceptions listed above, the server SHOULD respond with a 400 (Bad Request) response and close the connection.

Aside: the new holy scriptures are RFC9110 through RFC9114.  (They aren't _too different_ from RFC7230 and friends.)

**Comment by isomarcte** [comment]
@rossabaker I think that is handled here: https://github.com/http4s/http4s/blob/6931-special-case-handle-request-line-parse-failure/ember-server/shared/src/main/scala/org/http4s/ember/server/internal/ServerHelpers.scala#L504

e.g. when we hit a general `errorHandler` or now also `requestLineParseErrorHandler` case we yield `None` in the `fs2.Stream.unfold`, terminating the `Stream` and _I think_ closing the connection.

Does that line up with your understanding? I'm admit I'm relatively unfamiliar with the Ember internals.

**Comment by isomarcte** [comment]
@armanbilge @coltfred this is ready for review. I've added some tests.

**Comment by isomarcte** [comment]
@rossabaker I checked and this does close the connection. We get a RST back if we try to write to the client socket again after the bad request.

File: ember-server/shared/src/main/scala/org/http4s/ember/server/EmberServerBuilder.scala [hunk 2]
```diff
@@ -176,6 +179,23 @@ final class EmberServerBuilder[F[_]: Async: Network] private (
...
+  /** An error handler which will run in cases where the server is unable to
+    * parse the "start-line" (http 2 name) or "request-line" (http 1.1
+    * name). This is the first line of the request, e.g. "GET / HTTP/1.1".
+    *
+    * In this case, RFC 9112 (HTTP 2) says a 400 should be returned.
+    *
+    * This handler allows for configuring the behavior. The default as of
+    * 0.23.19 is to return a 400.
+    *
+    * @see [[https://www.rfc-editor.org/rfc/rfc9112#section-2.2-9 RFC 9112]
```
Comment (valencik) [inline ember-server/shared/src/main/scala/org/http4s/ember/server/EmberServerBuilder.scala:191]:
```suggestion
    * @see [[https://www.rfc-editor.org/rfc/rfc9112#section-2.2-9 RFC 9112]]
```

Fixes the scaladoc formatting

Comment (isomarcte) [reply ember-server/shared/src/main/scala/org/http4s/ember/server/EmberServerBuilder.scala]: Thanks @valencik. Updated.

**Comment by coltfred** [comment]
@valencik It looks like re-review would be nice. Any other followup here? It looks reasonable to me, but obviously I'm not super familiar with http4s internals.

**Review by danicheg — APPROVED** [review]
Despite I'm not expert in Ember internals, LGTM

**Comment by valencik** [comment]
Is the reason this is user configurable to allow the option for returning a 301, as per RFC 9112?

> Recipients of an invalid request-line SHOULD respond with either a 400 (Bad Request) error or a 301 (Moved Permanently) redirect with the request-target properly encoded. A recipient SHOULD NOT attempt to autocorrect and then process the request without a redirect, since the invalid request-line might be deliberately crafted to bypass security filters along the request chain.

It seems the RFC is suggesting you either return a 400 or a 301, but we're adding a user configurable handler that could any type of response.
Notably we're testing with a handler that returns a 500 which doesn't seem recommended in the RFC.

I guess I'm wondering (out loud) if this is more configuration than needed?
Perhaps it would have been ok to just add the extra error case on `ParsePreludeError` to return a 400, with no user configuration?

I'll admit it's not clear to me when/how it would be appropriate to return a 301 for an invalid request line. :thinking:

**Comment by coltfred** [comment]
The flexibility of allowing the caller to override the handler for this case seems good to me. It's defaulted to what we can all imagine, but if someone wants to override to do handling which results in a 301, they can. Making it a simple boolean doesn't allow for redirect specific behavior.

I'd love to get this merged so I can reduce the noise in my production logs. 😏

**Comment by valencik** [comment]
TLDR: I don't think I know enough to approve this PR, but I certainly wouldn't reject it. I think we need to tag someone else in.

The bug fix of returning a 400 on the bad parse is a clear win, would approve and merge. ✅ 
But the additional handler to the ember builder, for some reason seems like a lot.

Again, perhaps just because I don't really get why a 301 would be more appropriate than a 400.
If someone sends a request with `GET /hello/ember?foo={ HTTP/1.0\r\n\r\n` why is redirecting better than returning an error?
Also I get the impression that no one here actually intends to use the 301 option?
I mean, the spec says a 301 is allowed, so perhaps I shouldn't judge.

Either way, I'm new here and a bit conflicted, so I'm tagging in some experts in hopes they can shine some light.

@rossabaker @ChristopherDavenport

**Comment by armanbilge** [comment]
> The bug fix of returning a 400 on the bad parse is a clear win, would approve and merge. ✅
> But the additional handler to the ember builder, for some reason seems like a lot.

FWIW I've been feeling the same.

**Comment by coltfred** [comment]
As a consumer of http4s I have had situations where I'm trying to implement edge cases of an RFC before and so even if we don't understand the redirect use case it seems important to support it.

I agree that the `Throwable => F[Response[F]]` is a bit more powerful than it needs to be, but off the top of my head I don't have a better suggestion.

**Comment by coltfred** [comment]
I hate to keep pushing here, but what can I do to move this forward? Is there an alternate abstraction that we think makes more sense?

**Comment by armanbilge** [comment]
> Is there an alternate abstraction that we think makes more sense?

It seems like there are currently exactly two behaviors which follow the spec. So instead of a `Throwable => F[Response[F]]` all we need is an enum with two cases.

**Comment by valencik** [comment]
Alternatively a PR that just returns the 400 on a bad parse and doesn't attempt to offer the new handler would be fine. We are currently out of spec, and such a bug fix would bring us back in line with the spec.

**Comment by coltfred** [comment]
The case of the 301 could need to decision where to redirect based on the error though, right? If that's the case I think we need to allow control of the full response because setting headers in addition of where to redirect.

I would be fine with what @valencik suggested where it returns a 400 with the possibility of adding configuration in the future.

**Comment by armanbilge** [comment]
> The case of the 301 could need to decision where to redirect based on the error though, right?

Hm, not sure. My understanding was it just re-directs to the properly encoded URI.

**Comment by isomarcte** [comment]
> Hm, not sure. My understanding was it just re-directs to the properly encoded URI.

Yes, that is what the spec does. 

I can update the PR to just do the 400 behavior for now and if anyone wants/needs/has time to write the 301 based behavior, then we can add a configuration option for that.

Does that sound like a path forward?

**Comment by coltfred** [comment]
That sounds awesome!

**Comment by armanbilge** [comment]
Sounds good, thanks!

**Comment by isomarcte** [comment]
Oh wait, before I do that, I recall now why I chose to expose this function. 

Because this failure occurs _before_ we construct a `org.http4s.Request[F]` value, it is obviously not visible in the `HttpApp`/`HttpRoutes` being run by ember. 

`Throwable => Response[F]` allows a hook for observing the failure, for logs or whatever, not just changing the response. It seemed like we should let this be observable. I could provide a function `Throwable => F[Unit]`, but at the time that `Throwable => F[Response[F]]` with a default seemed better. 

Would you guys prefer I hard code the 400 behavior with no observability? Add the `Throwable => F[Unit]` hook? Or keep it `Throwable => F[Response[F]]`.

I'm not personally hitting this issue in prod, but I know that if I was I'd definitely want to be able to hook whatever metrics/logs/etc into this.

**Comment by armanbilge** [comment]
Aha, that's a helpful clarification! So basically two things got conflated.
1. Configuring the `Response` for unparseable request lines.
2. Configuring an error-hook for observing these sorts of failures.

These are both good things, but they very much seem like they should be separate? Particularly since for (1) there are only two specific options prescribed by the spec, but for (2) it can be anything the user wants.

**Comment by isomarcte** [comment]
Sounds like you'd like `Throwable = F[Unit]`? :wink:

**Comment by isomarcte** [comment]
I can make that a separate PR. Probably should be called for the normal `errorHandler` too then.

**Comment by coltfred** [comment]
I personally like the `Throwable => F[Response[F]]` since it allows the http4s server writer the ability to send back a body that is meaningful for their API. Many people have a standard "error response" format which the `Throwable => F[Unit]` does not provide.

At this point I'll take whatever we can agree on, but I think if there is going to be something in the Builder it might as well be flexible enough to allow API writers to do what they need.

**Comment by isomarcte** [comment]
If we wanted to support @coltfred's use case and we want to only have the handler yield the body (which would imply some headers for encoding) then things start to get gross...something like `Throwable => F[(Entity[F], NonEmptyChain[Header])]` or something...

Perhaps a compromise alternative could be we leave it at `Throwable => F[Response[F]]`, but we check the status code and rewrite it to 400 if the user code doesn't set it to 301 or 400? I don't love it, because if it ever triggers it's going to be super confusing what's going on, but OTOH I doubt many people will be fooling with this.

Or we just leave it as `Throwable => F[Response[F]]` with a valid default and if the user breaks the spec, then they break the spec. Users can already break the spec by creating nonsensical request/response instances, like a `HEAD` request with a body, AFAIK.

**Comment by armanbilge** [comment]
> Or we just leave it as `Throwable => F[Response[F]]` with a valid default and if the user breaks the spec, then they break the spec.

Ok, I see the reasoning now. I think I'm convinced, but would be good to hear from others as well. Thanks for your patience :)

File: ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala [hunk 3]
```diff
@@ -165,4 +179,34 @@ class ConnectionSuite extends Http4sSuite {
...
+    for {
+      _ <- client.writes(fs2.text.utf8.encode(request))
+      resp <- client.responsePrelude
+    } yield assertEquals(resp.status, Status.BadRequest)
+  }
+
+  ResourceFunFixture(
+    emberServerBuilder(defaultIdleTimeout, defaultHeaderTimeout)
+      .withRequestLineParseErrorHandler { case _ =>
+        IO.pure(Response(Status.InternalServerError))
```
Comment (armanbilge) [inline ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: Feels naughty to test something that shouldn't be allowed by spec 😉

**Review by armanbilge — APPROVED** [review]

File: ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala (hunk 3 above)
Comment (isomarcte) [reply ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: haha. I'll change it to 301

Comment (coltfred) [reply ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: @isomarcte Do you want to fix that test?

Comment (armanbilge) [reply ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: It's not a big deal, let's not block on this any longer.

Comment (coltfred) [reply ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: Can we merge, then? Did @rossabaker or @ChristopherDavenport want to chime in at all on this?

Comment (armanbilge) [reply ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: Yeah, I'm hoping for one more look at this :)

**Review by ChristopherDavenport — APPROVED** [review]

**Review by valencik — APPROVED** [review]
Thanks for the patience and all the extra context folks.

File: ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala (hunk 3 above)
Comment (isomarcte) [reply ember-server/shared/src/test/scala/org/http4s/ember/server/ConnectionSuite.scala:196]: Sorry, @armanbilge @coltfred, I've been short on time recently. I can update that test in another PR.
